### ISSUE TITLE
8.1 Build a calculator (Optional Pairing)

### ISSUE THREAD

**Issue by engie15** [phase-0/phase-0#4539]
# 8.1 Build a calculator (Optional Pairing)
Work on the [Free Code Camp 'Build a calculator'](http://www.freecodecamp.com/challenges/build-a-javascript-calculator) project, ideally with another member of your targetted bootcamp cohort. Keep it simple, it's well within your capabilities. And *don't look at the sample solution yet*.

**[With your pair](https://github.com/dev-academy-programme/curriculum/tree/master/concepts/pairing)** *if possible*:

- [x] Start Toggl timer.
- [x] If you have one, check-in with your pair:
  - How did you sleep?
  - How do you feel now?
  - Is there anything going on in your life that might affect how you are working right now?
- [x] Create a new git repository called "js-calculator".
- [x] Reverse pseudocode the [sample solution](http://codepen.io/FreeCodeCamp/pen/EPNZYW) in a waffle comment below. *Remember, there is no 'right way' - the sample solution is just one implementation*.
- [x] Build the app:
  - [x] Start with building the html, think about what buttons are needed on a calculator
  - [x] Next, give your calculator some style. Use skeleton, raw CSS or whatever you feel like to make your buttons look more like a real calculator.
  - [x] Finally, start working through the Javascript, this is the hardest part, so...
    - Take your time and work through small problems.
    - Make small commits
    - Refer to your pseudocode.
    - Use the Developer Tools in your browser.
    - Reference previous projects/materials for app structure
    - (If pairing, Change driver-navigator roles approx. every 30 mins)
    - Ask for help. You can do this!
- [x] If you did this challenge as a pair, spend 10 minutes asking each other these questions:
  - Make sure your feedback is ASK (actionable, specific, kind)
  - [x] What did you find awesome about our pairing session?
  - [x] What could have been more awesome?
    - *This question and answer may feel difficult, but is a very valuable learning tool. It can be kind to gently help someone identify their weaknesses so that they can work on them. Remember too that your pair may not completely understand the concept in question, so take their feedback [with a grain of salt](http://www.phrases.org.uk/meanings/take-with-a-grain-of-salt.html).*
  - [x] Did you notice any gaps in my understanding?

**Comment by lauren-hart** [comment]
Declare a variable - entries as an empty array []
var entries = [];

Declare a variable - total and set it to zero.
var total = 0;

Declare a variable - temp and set it to an empty string.
var temp = ' ';

Using jQuery - create an on click event where anything with
the element “button” will run a function which will create a 
new variable - val. It will then set the new variable - val to be 
the text of the button pushed. 

$("button").on('click', function() {
 	var val = $(this).text();

Then an if statement is run, it is saying if the value is a number or the
value is equal to a decimal then it will add the val (value) to the empty string (temp).

  // Got a number, add to temp
  if (!isNaN(val) || val === '.') {
    temp += val;

We then run another jQuery function where we look for anything with the ID of answer and the value is assigned to the temp string by using the substring function. 

 $("#answer").val(temp.substring(0,10));

By specifying 0 and 10, we are saying we want the string to start from position 0 and we want to show up to position 10. So anything that has more decimal places than 10 will not be displayed. 

The if statement is continued and it is now saying else if (it isn’t a number or decimal) is it equal to AC? If it is, then entries will be reset to an empty array, temp will be reset to an empty string, the total will be reset to 0. The jQuery function than sets the id answer (display screen) as the value of ‘ ‘ (an empty string). 

  // Got some symbol other than equals, add temp to our entries
  // then add our current symbol and clear temp
  } else if (val === 'AC') {
    entries = [];
    temp = '';
    total = 0;
    $("#answer").val('')

We then keep running the if statement, so if it isn’t a number and it isn’t equal to AC but it is equal to CE, we then set the temp string to ‘ ‘ (empty string. It is different to above as we are not clearing the entries array, or resetting the total to zero, we are only changing the temp string. We then push the val(‘ ‘) to the display screen. 

  // Clear last entry
  } else if (val === 'CE') {
    temp = '';
    $("#answer").val('')

We then keep running the if statement, so if it isn’t a number and it isn’t equal to AC and it isn’t equal to CE but it is equal to ‘x’, we use the function push to push the temp string to the entries array. We then set the temp string to * which is what we are pushing to the entries array. 

// Change multiply symbol to work with eval
  } else if (val === 'x') {
    entries.push(temp);
    entries.push('*');
    temp = ' ';

We then keep running the if statement, so if it isn’t a number and it isn’t equal to AC and it isn’t equal to CE, and it isn’t equal to x but it is equal to divide we do the same thing as above but we set the ‘/‘ symbol as the temp string to push to the empty array.

  // Change divide symbol to work with eval
  } else if (val === '÷') {
    entries.push(temp);
    entries.push('/');
    temp = '';


We then keep running the if statement, so if it isn’t a number and it isn’t equal to AC and it isn’t equal to CE, and it isn’t equal to x and it isn’t equal to divide but it is equal to ‘=‘ we go through the following.

First we push temp to the array entries.

// Got the equals sign, perform calculation
  } else if (val === '=') {
  	entries.push(temp);

We set a variable “nt” which is number(entries[0]). 

The number function converts the object argument to a number. So nt is converting the entries string (at index 0) to a number.

    var nt = Number(entries[0]);

We then run a for loop, where we are looping through the length of the entries array. 

A new variable is declared nextNum and is set as the number(entries[i+1]
A new variable symbol is declared and is set as entries[i]


    for (var i = 1; i < entries.length; i++) {
      var nextNum = Number(entries[i+1])
      var symbol = entries[i];

We then run an if statement, the first argument is if the symbol is equal to + then we add the number (var nt) to the next number (nextNum), we then do the same for subtract, multiply and divide. 
      
      if (symbol === '+') { nt += nextNum; } 
      else if (symbol === '-') { nt -= nextNum; } 
      else if (symbol === '*') { nt *= nextNum; } 
      else if (symbol === '/') { nt /= nextNum; }
      
      i++;
    }

To set the number as a negative if it is a negative we run an if statement to see if it is less than zero and than we assign nt as the math.abs and add the negative symbol. 
    
    // Swap the '-' symbol so text input handles it correctly
    if (nt < 0) {
      nt = Math.abs(nt) + '-';
    }
    




    $("#answer").val(nt);
		entries = [];
    temp = '';
    
  // Push number
  } else {
    entries.push(temp);
    entries.push(val);
    temp = '';
  }
});
